src/enzo/Grid_AccreteOntoAccretingParticle.C

```diff
@@ -0,0 +1,477 @@
+
+	  else
+	  {
+	    // Keep cell mass well above density floor
+	    if ((mcell - maccreted)/CellVolume > SmallRhoFac*SmallRho)
+	    {
+	      mnew[index] = mcell - maccreted;
+	    }
+	    else
+	    {
+	      mnew[index] = SmallRhoFac*SmallRho*CellVolume;
+	      maccreted = mcell - mnew[index];
+	    }
+
+	    rhonew = mnew[index]/CellVolume;
+
+	    // Find the radius vector
+	    reff[0] = (CellLeftEdge[0][i] + 0.5*CellWidth[0][i]) - xsink;
+	    reff[1] = (CellLeftEdge[1][j] + 0.5*CellWidth[1][j]) - ysink;
+	    reff[2] = (CellLeftEdge[2][k] + 0.5*CellWidth[2][k]) - zsink;
+	    rsqr = reff[0]*reff[0]+reff[1]*reff[1]+reff[2]*reff[2];
+
+	    // Prevent a floating point error if close to central cell
+	    if (rsqr <= POW(1e-7*CellWidth[0][0],2))
+	    {
+	      reff[0] = 0.0;
+	      reff[1] = 0.0;
+	      reff[2] = 0.0;
+	      rsqr = 1.0;
+	    }
+
+	    // Compute the amount of momentum accreted
+	    paccrete[0][index] = pcell[0] * maccreted/mcell;
+	    paccrete[1][index] = pcell[1] * maccreted/mcell;
+	    paccrete[2][index] = pcell[2] * maccreted/mcell;
+
+	    // Compute new total internal energy. By construction,
+	    // this keeps the specific internal energy constant after
+	    // accretion
+	    eintnew = eint * (1.0 - maccreted/mcell);
+
+	    /* Compute the new momentum of the cell.  Note that we do
+	       not use pcell here because we need to do this
+	       calculation in the grid frame, not the sink frame. */
+	    pnew[0][index] = mcell*vgas[0] - paccrete[0][index];
+	    pnew[1][index] = mcell*vgas[1] - paccrete[1][index];
+	    pnew[2][index] = mcell*vgas[2] - paccrete[2][index];
+
+	    // Compute new total kinetic energy (not density)
+	    kenew = (pnew[0][index]*pnew[0][index] +
+		     pnew[1][index]*pnew[1][index] +
+		     pnew[2][index]*pnew[2][index]) /
+	            (2.0 * mnew[index]);
+
+	    // Compute the new total energy
+	    etotnew = eintnew + kenew;
+
+	    // This is actually a density since particle masses are stored
+	    // in density units.
+	    AccretedMass += maccreted/CellVolume;
+	    AccretedMomentum[0] += paccrete[0][index];
+	    AccretedMomentum[1] += paccrete[1][index];
+	    AccretedMomentum[2] += paccrete[2][index];
+
+#ifdef DEBUG_AP
+	    if (index == cgindex)
+	      printf("Sink Density: %"GOUTSYM", "
+		     "Cell Density: %"GOUTSYM", "
+		     "New Density:  %"GOUTSYM"\n",
+		     maccreted/CellVolume,
+		     BaryonField[DensNum][index],
+		     BaryonField[DensNum][index]-maccreted/CellVolume);
+#endif
+
+	    // Update the density
+	    BaryonField[DensNum][index] -= maccreted/CellVolume;
+
+	    // Update the grid velocities
+	    if (HydroMethod == PPM_DirectEuler)
+	    {
+	      BaryonField[Vel1Num][index] = pnew[0][index]/mnew[index];
+	      BaryonField[Vel2Num][index] = pnew[1][index]/mnew[index];
+	      BaryonField[Vel3Num][index] = pnew[2][index]/mnew[index];
+	    }
+	    else if (HydroMethod == Zeus_Hydro)
+	    {
+	      // do nothing for now.
+	    }
```

Comment:
Would be useful to  add the upgrade of the BaryonField for Zeus_Hydro method before merging.

Reply:
The accreting particle algorithm has a bug that causes the particle to not conserve momentum. I was able to show that it breaks with the Burkert-Bodenheimer test. I think it's best to leave it for now though as it was the original inspiration for the SmartStar particle and contains a lot of the underlying physics. 

At the moment the only AP I would trust is the SmartStar since it's the only one that has been extensively used/tested. All the other particle types are shells until someone tests/uses them.